File: adventofcode-go/solutions/day1.go
package main

import "fmt"
import "github.com/adokarasev/adventofcode2019/adventofcode-go/utils"

func main() {
	input, _ := utils.ReadInts("../input/day1.txt")
	fmt.Printf("1. Total fuel: %v\n", sum(input, calculateFuel))
	fmt.Printf("2. Total fuel: %v\n", sum(input, calculateRequiredFuel))
}

func sum(input []int, f func(int) int) int {
	total := 0
	for _, m := range input {
		total += f(m)
	}
	return total
}

func calculateFuel(m int) int {
	return m/3 - 2
}

func calculateRequiredFuel(m int) int {
	f := m/3 - 2
	if f < 0 {
		return 0
	}
	return f + calculateRequiredFuel(f)
}

File: adventofcode-kotlin/src/main/kotlin/com/adventofcode/Day1.kt
package com.adventofcode

fun main() {
    val input = Utils.readInts("../input/day1.txt")

    input
        .fold(0) { acc, c -> acc + c / 3 - 2 }
        .let { println("1. Total fuel: $it") }

    input
        .fold(0) { acc, c -> acc + calculateRequiredFuel(c) }
        .let { println("2. Total fuel: $it") }
}

fun calculateRequiredFuel(m: Int): Int {
    val f = m / 3 - 2
    return if (f < 0) 0 else f + calculateRequiredFuel(f)
}
